src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java

```diff
@@ -170,31 +171,69 @@ public void close() {
     ) {
         // if it's a starting clause
         if (this._child == null) {
-            // create initial RDD from expression
-            JavaRDD<Item> initialRdd = _expression.getRDD(context);
+            return getDataFrameFromRDDExpression(context);
+        }
 
-            // define a schema
-            List<StructField> fields = new ArrayList<>();
-            StructField field = DataTypes.createStructField(_variableName, DataTypes.BinaryType, true);
-            fields.add(field);
-            StructType schema = DataTypes.createStructType(fields);
+        if (_child.isDataFrame()) {
+            if (_expression.isRDD()) {
+                Set<String> intersection = new HashSet<>(_expression.getVariableDependencies().keySet());
+                intersection.retainAll(getVariablesBoundInCurrentFLWORExpression());
+                boolean expressionUsesVariablesOfCurrentFlwor = !intersection.isEmpty();
+
+                if (expressionUsesVariablesOfCurrentFlwor) {
+                    throw new InvalidForClauseExpressionException(
+                            "An RDD-enabled for clause expression cannot reference the variables defined in the encapsulating, RDD-enabled FLWOR expression.",
+                            getMetadata().getExpressionMetadata()
+                    );
+                }
 
-            JavaRDD<Row> rowRDD = initialRdd.map(new ForClauseSerializeClosure());
+                // since no variable dependency to the current FLWOR expression exists for the expression
+                // evaluate the DF with the parent context and calculate the cartesian product
+                Dataset<Row> expressionDF;
+                if (_expression.isRDD()) {
```

Comment:
We already know that the expression is an RDD at this point, so that the test is superfluous and the else clause can go away.

Reply:
I had missed this when moving the code around. Thanks for catching this.